src/plugin/src/main/java/com/defenseunicorns/uds/keycloak/plugin/authentication/RequireGroupAuthenticator.java

```diff
@@ -0,0 +1,144 @@
+package com.defenseunicorns.uds.keycloak.plugin.authentication;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * Simple {@link Authenticator} that checks if a user is a member of a given {@link GroupModel Group}.
+ */
+public class RequireGroupAuthenticator implements Authenticator {
+
+    private static final Logger LOGGER = Logger.getLogger(RequireGroupAuthenticator.class.getName());
+
+    @Override
+    public void authenticate(final AuthenticationFlowContext context) {
+        UserModel user = context.getUser();
+        if (user == null) {
+            LOGGER.warn("Invalid user");
+            context.failure(AuthenticationFlowError.INVALID_USER);
+            return;
+        }
+
+        ClientModel client = context.getAuthenticationSession().getClient();
+        String groupsAttribute = client.getAttribute("uds.core.groups");
+
+        if (groupsAttribute == null || groupsAttribute.isEmpty()) {
+            LOGGER.infof("No groups detected for client %s", client.getName());
+            context.success();
+            return;
+        }
+
+        Groups groups = parseGroupsAttribute(groupsAttribute);
+        if (groups == null) {
+            LOGGER.warn("Failed to parse groups JSON");
+            context.failure(AuthenticationFlowError.INVALID_CLIENT_SESSION);
+            return;
+        }
+
+        RealmModel realm = context.getRealm();
+        boolean foundGroup = false;
+        List<String> requiredGroups = new ArrayList<>();
+
+        for (String groupName : groups.anyOf) {
+            requiredGroups.add(groupName);
+            Optional<GroupModel> group = getGroupByPath(groupName, realm);
+
+            if (group.isPresent()) {
+                if (isMemberOfExactGroup(user, group.get())) {
+                    LOGGER.infof("User %s is authorized for group %s", user.getUsername(), groupName);
+                    context.success();
+                    return;
+                }
+                foundGroup = true;
+            }
+        }
+
+        if (foundGroup) {
+            LOGGER.warn("User is not a member of the required group(s)");
+            context.failure(AuthenticationFlowError.INVALID_CLIENT_SESSION);
+            return;
+        }
+        LOGGER.warnf("Required group(s) %s do not exist in realm", requiredGroups);
```

Comment:
After switching to a `Groups` object, remove this ArrayList (and remove it's `import` too)
```suggestion

        for (String groupName : groups.anyOf) {
            requiredGroups.add(groupName);
            Optional<GroupModel> group = getGroupByPath(groupName, realm);

            if (group.isPresent()) {
                if (isMemberOfExactGroup(user, group.get())) {
                    LOGGER.infof("User %s is authorized for group %s", user.getUsername(), groupName);
                    context.success();
                    return;
                }
                foundGroup = true;
            }
        }

        if (foundGroup) {
            LOGGER.warn("User is not a member of the required group(s)");
            context.failure(AuthenticationFlowError.INVALID_CLIENT_SESSION);
            return;
        }
        LOGGER.warnf("Required group(s) %s do not exist in realm", Arrays.toString(groups.anyOf)));
```

Reply:
updated